File: Pharmaceutical_Company_Example/01_Creational/Python/Abstract_Factory.py
from abc import ABC, abstractmethod

# Abstract Factory Interface
class RegulatoryFactory(ABC):
    @abstractmethod
    def create_submission_form(self):
        pass

    @abstractmethod
    def create_testing_protocol(self):
        pass

# Concrete Factory for Vaccines
class VaccineRegulatoryFactory(RegulatoryFactory):
    def create_submission_form(self):
        return VaccineSubmissionForm()

    def create_testing_protocol(self):
        return VaccineTestingProtocol()

# Concrete Factory for Small-Molecule Drugs
class SmallMoleculeRegulatoryFactory(RegulatoryFactory):
    def create_submission_form(self):
        return SmallMoleculeSubmissionForm()

    def create_testing_protocol(self):
        return SmallMoleculeTestingProtocol()

# Concrete Factory for Biologics
class BiologicsRegulatoryFactory(RegulatoryFactory):
    def create_submission_form(self):
        return BiologicsSubmissionForm()

    def create_testing_protocol(self):
        return BiologicsTestingProtocol()

# Abstract Product
class SubmissionForm(ABC):
    @abstractmethod
    def submit(self):
        pass

# Concrete Product: Vaccine Submission Form
class VaccineSubmissionForm(SubmissionForm):
    def submit(self):
        print("Submitting vaccine clinical trial data form.")

# Concrete Product: Small-Molecule Submission Form
class SmallMoleculeSubmissionForm(SubmissionForm):
    def submit(self):
        print("Submitting small-molecule drug trial data form.")

# Concrete Product: Biologics Submission Form
class BiologicsSubmissionForm(SubmissionForm):
    def submit(self):
        print("Submitting biologics clinical trial data form.")

# Abstract Product
class TestingProtocol(ABC):
    @abstractmethod
    def run_tests(self):
        pass

# Concrete Product: Vaccine Testing Protocol
class VaccineTestingProtocol(TestingProtocol):
    def run_tests(self):
        print("Running tests for vaccines.")

# Concrete Product: Small-Molecule Testing Protocol
class SmallMoleculeTestingProtocol(TestingProtocol):
    def run_tests(self):
        print("Running tests for small-molecule drugs.")

# Concrete Product: Biologics Testing Protocol
class BiologicsTestingProtocol(TestingProtocol):
    def run_tests(self):
        print("Running tests for biologics.")


def process_regulatory_approval(factory: RegulatoryFactory):
    submission_form = factory.create_submission_form()
    testing_protocol = factory.create_testing_protocol()
    
    submission_form.submit()
    testing_protocol.run_tests()

# Vaccine Regulatory Process
vaccine_factory = VaccineRegulatoryFactory()
process_regulatory_approval(vaccine_factory)

# Small-Molecule Drug Regulatory Process
small_molecule_factory = SmallMoleculeRegulatoryFactory()
process_regulatory_approval(small_molecule_factory)

# Biologics Regulatory Process
biologics_factory = BiologicsRegulatoryFactory()
process_regulatory_approval(biologics_factory)

File: Pharmaceutical_Company_Example/01_Creational/PHP/Abstract_Factory.php
<?php

// Abstract Factory Interface
interface RegulatoryFactory {
    public function createSubmissionForm(): SubmissionForm;
    public function createTestingProtocol(): TestingProtocol;
}

// Concrete Factory for Vaccines
class VaccineRegulatoryFactory implements RegulatoryFactory {
    public function createSubmissionForm(): SubmissionForm {
        return new VaccineSubmissionForm();
    }

    public function createTestingProtocol(): TestingProtocol {
        return new VaccineTestingProtocol();
    }
}

// Concrete Factory for Small-Molecule Drugs
class SmallMoleculeRegulatoryFactory implements RegulatoryFactory {
    public function createSubmissionForm(): SubmissionForm {
        return new SmallMoleculeSubmissionForm();
    }

    public function createTestingProtocol(): TestingProtocol {
        return new SmallMoleculeTestingProtocol();
    }
}

// Concrete Factory for Biologics
class BiologicsRegulatoryFactory implements RegulatoryFactory {
    public function createSubmissionForm(): SubmissionForm {
        return new BiologicsSubmissionForm();
    }

    public function createTestingProtocol(): TestingProtocol {
        return new BiologicsTestingProtocol();
    }
}

// Abstract Product Interface for Submission Forms
interface SubmissionForm {
    public function submit(): void;
}

// Concrete Product: Vaccine Submission Form
class VaccineSubmissionForm implements SubmissionForm {
    public function submit(): void {
        echo "Submitting vaccine clinical trial data form.\n";
    }
}

// Concrete Product: Small-Molecule Submission Form
class SmallMoleculeSubmissionForm implements SubmissionForm {
    public function submit(): void {
        echo "Submitting small-molecule drug trial data form.\n";
    }
}

// Concrete Product: Biologics Submission Form
class BiologicsSubmissionForm implements SubmissionForm {
    public function submit(): void {
        echo "Submitting biologics clinical trial data form.\n";
    }
}

// Abstract Product Interface for Testing Protocols
interface TestingProtocol {
    public function runTests(): void;
}

// Concrete Product: Vaccine Testing Protocol
class VaccineTestingProtocol implements TestingProtocol {
    public function runTests(): void {
        echo "Running tests for vaccines.\n";
    }
}

// Concrete Product: Small-Molecule Testing Protocol
class SmallMoleculeTestingProtocol implements TestingProtocol {
    public function runTests(): void {
        echo "Running tests for small-molecule drugs.\n";
    }
}

// Concrete Product: Biologics Testing Protocol
class BiologicsTestingProtocol implements TestingProtocol {
    public function runTests(): void {
        echo "Running tests for biologics.\n";
    }
}

// Function to Process Regulatory Approval
function processRegulatoryApproval(RegulatoryFactory $factory): void {
    $submissionForm = $factory->createSubmissionForm();
    $testingProtocol = $factory->createTestingProtocol();
    
    $submissionForm->submit();
    $testingProtocol->runTests();
}

// Vaccine Regulatory Process
$vaccineFactory = new VaccineRegulatoryFactory();
processRegulatoryApproval($vaccineFactory);

// Small-Molecule Drug Regulatory Process
$smallMoleculeFactory = new SmallMoleculeRegulatoryFactory();
processRegulatoryApproval($smallMoleculeFactory);

// Biologics Regulatory Process
$biologicsFactory = new BiologicsRegulatoryFactory();
processRegulatoryApproval($biologicsFactory);

?>
